mempool/reactor.go

```diff
@@ -317,3 +434,216 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
 		}
 	}
 }
+
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating
+	// transactions, where source and target are node IDs.
+	disabledRoutes map[p2p.ID]map[p2p.ID]struct{}
+}
+
+func newGossipRouter() *gossipRouter {
+	return &gossipRouter{
+		disabledRoutes: make(map[p2p.ID]map[p2p.ID]struct{}),
+	}
+}
+
+// disableRoute marks the route `source -> target` as disabled.
+func (r *gossipRouter) disableRoute(source, target p2p.ID) {
+	if source == noSender || target == noSender {
+		// TODO: this shouldn't happen
+		return
+	}
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	targets, ok := r.disabledRoutes[source]
+	if !ok {
+		targets = make(map[p2p.ID]struct{})
+	}
+	targets[target] = struct{}{}
+	r.disabledRoutes[source] = targets
+}
+
+// isRouteEnabled returns true iff the route source->target is disabled.
+func (r *gossipRouter) isRouteDisabled(source, target p2p.ID) bool {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+	if targets, ok := r.disabledRoutes[source]; ok {
+		if _, ok := targets[target]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// resetRoutes removes all disabled routes with peerID as source or target.
+func (r *gossipRouter) resetRoutes(peerID p2p.ID) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	// Remove peer as source.
+	delete(r.disabledRoutes, peerID)
+
+	// Remove peer as target.
+	for _, targets := range r.disabledRoutes {
+		delete(targets, peerID)
+	}
+}
+
+// resetRandomRouteWithTarget removes a random disabled route that has the given
+// target.
+func (r *gossipRouter) resetRandomRouteWithTarget(target p2p.ID) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	sourcesWithTarget := make([]p2p.ID, 0)
+	for s, targets := range r.disabledRoutes {
+		if _, ok := targets[target]; ok {
+			sourcesWithTarget = append(sourcesWithTarget, s)
+		}
+	}
+
+	if len(sourcesWithTarget) > 0 {
+		randomSource := sourcesWithTarget[cmtrand.Intn(len(sourcesWithTarget))]
+		delete(r.disabledRoutes[randomSource], target)
```

Comment:
To check. Shouldn't we call `resetRoutes(randomSource)` here?

Reply:
No. `resetRoutes(peerID)` removes all routes with `peerID` as source or target. Here we just want to remove one route.